IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.m

```diff
@@ -0,0 +1,208 @@
+
+@implementation MSIDMacCredentialStorageItem
+
+- (instancetype)init
+{
+    if (self = [super init])
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.universalstorage-%@", [NSUUID UUID].UUIDString];
+        self.queue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
+- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        NSString *credentialKey = [NSString stringWithFormat:@"%@%@%@", key.account, keyDelimiter, key.service];
+        [self.cacheObjects setObject:credential forKey:credentialKey];
+    });
+}
+
+- (void)mergeStorageItem:(MSIDMacCredentialStorageItem *)storageItem
+{
+    dispatch_barrier_async(self.queue, ^{
+        for (NSString *key in storageItem.cacheObjects)
+        {
+            MSIDCredentialCacheItem *credential = [storageItem.cacheObjects objectForKey:key];
+            if (credential)
+            {
+                [self.cacheObjects setObject:credential forKey:key];
+            }
+        }
+    });
+}
+
+- (void)removeStoredCredentialForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        NSString *credentialKey = [NSString stringWithFormat:@"%@%@%@", key.account, keyDelimiter, key.service];
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelInfo, nil, @"Removing credential for key %@.", MSID_PII_LOG_MASKABLE(credentialKey));
+        [self.cacheObjects removeObjectForKey:credentialKey];
+    });
+}
+
+- (NSArray<MSIDCredentialCacheItem *> *)storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key
+{
+    __block NSArray<MSIDCredentialCacheItem *> *credentials =  @[];
+    
+    dispatch_sync(self.queue, ^{
+        if (key.account && key.service)
+        {
+            NSString *credentialKey = [NSString stringWithFormat:@"%@%@%@", key.account, keyDelimiter, key.service];
+            MSIDCredentialCacheItem *credential = [self.cacheObjects objectForKey:credentialKey];
+            if (credential)
+            {
+                credentials = @[credential];
+            }
+        }
+        else
+        {
+            NSArray *storedCredentials = [self.cacheObjects allValues];
+            credentials = [storedCredentials filteredArrayUsingPredicate:[self createPredicateForKey:key]];
+        }
+    });
+    
+    return credentials;
+}
+
+/*
+ This api is thread safe only if an immutable object is passed as parameter.
+ */
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json
+                                 error:(NSError * __autoreleasing *)error
+{
+    MSIDMacCredentialStorageItem *instance = [self init];
+    
+    if (instance)
+    {
+        for (NSString *credentialKey in json)
+        {
+            NSDictionary *rtDict = [json objectForKey:credentialKey];
```

Comment:
you can use the helper we have "msidObjectForKey:class:" or something similar that does the check for you in a consistent manner

Reply:
Sure. Updated to use the helper method.